internal/terraform/node_resource_validate.go

```diff
@@ -383,10 +384,26 @@ func (n *NodeValidatableResource) validateResource(ctx EvalContext) tfdiags.Diag
 				moreDiags := schema.StaticValidateTraversal(traversal)
 				diags = diags.Append(moreDiags)
 
-				// TODO: we want to notify users that they can't use
-				// ignore_changes for computed attributes, but we don't have an
-				// easy way to correlate the config value, schema and
-				// traversal together.
+				// ignore_changes cannot be used for Computed attributes,
+				// unless they are also Optional.
+				// If the traversal was valid, convert it to a cty.Path and
+				// use that to check whether the Attribute is Computed and
+				// non-Optional.
+				if !diags.HasErrors() {
+					path := traversalToPath(traversal)
+
+					attrSchema := schema.AttributeByPath(path)
+
+					if attrSchema != nil && !attrSchema.Optional && attrSchema.Computed {
+						diags = diags.Append(&hcl.Diagnostic{
+							Severity: hcl.DiagWarning,
+							Summary:  "Invalid ignore_changes",
+							Detail:   fmt.Sprintf("ignore_changes cannot be used with non-Optional Computed attributes. Please remove \"%s\" from ignore_changes.", hcl2shim.FlatmapKeyFromPath(path)),
```

Comment:
`FlatmapKeyFromPath` is not going to create correct hcl2 strings for indexed paths (plus we've managed to remove `hcl2shim` from all non-test core code). Let's see if we can find an existing formatting method or make abetter one -- I know I've needed one recently too.

Reply:
Ah, found it, `tfdiags.FormatCtyPath`

Reply:
`tfdiags.FormatCtyPath()` is widely used, but outputs `.computed_string`, which I think would be more helpful without the initial `.`. 

Would it be better to use `traversalStr()` (perhaps moved somewhere less specific), or write a new method? Or just use `FormatCtyPath` for consistency?

Reply:
I guess what you've encountered here is a follow-on consequence of the existing inconsistency that `ignore_changes` takes what HCL calls "relative traversals" but does so using the _absolute traversal_ syntax, and so in the flat single-attribute case it's just a single attribute name without any extra punctuation.

The `tfdiags.FormatCtyPath` function is typically used to report evaluation errors where a problem occurs in some deep location inside a data structure and where the evaluator only reports the relative traversal from the perspective of the expression that failed, and so the leading `.` makes sense in that case to help explain that it's talking about a relative path rather than an absolute one. :confounded: 

I can't think of any other situation where we currently expect a user to write a relative traversal using absolute traversal syntax, but I know you're working on something that may introduce a second example! :grinning: Given the relatively narrow use of this though, I wonder about being a bit pragmatic here and perhaps just testing if the result from `FormatCtyPath` has a leading `.` and trimming it off if so; this seems unlikely to arise in any other situations and so I hesitate to add another traversal/path formatter just to handle this edge case. :thinking:

Reply:
Thanks, I agree this is pragmatic for now. Once I introduce the "second example" of this formatting I'll propose another formatting function, as it'll no longer be an edge case.